File: ruby/meetup/meetup.rb
require 'date'

# Not sure if monkey patching Date is a good practice here
class Date
  def first?
    day >= 1 && day <= 7
  end

  def second?
    day >= 8 && day <= 14
  end

  def third?
    day >= 15 && day <= 21
  end

  def fourth?
    day >= 22 && day <= 28
  end

  def teenth?
    day >= 13 && day <= 19
  end

  def last?
    day >= 23 && day <= 31
  end
end

# Meetup
class Meetup
  attr_accessor :month, :year, :weekday, :schedule
  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, schedule)
    @weekday = weekday
    @schedule = schedule
    find_meetup_date
  end

  private

  # Start with last date in month for 'last' case
  def find_meetup_date
    Date.new(year, month).next_month.prev_day
        .downto(Date.new(year, month)) do |date|
      return date if correct_date?(date)
    end
  end

  def correct_date?(date)
    correct_weekday?(date) && correct_schedule?(date)
  end

  def correct_weekday?(date)
    date.send(weekday.to_s + '?')
  end

  def correct_schedule?(date)
    date.send(schedule.to_s + '?')
  end
end

File: python/meetup/meetup.py
from datetime import date
from calendar import Calendar

def meetup_day(year, month, weekday, schedule):
    candidates = [date for date in Calendar().itermonthdates(year, month)
                    if date.month == month and \
                    date.strftime('%A') == weekday ]

    if schedule == 'teenth':
        return find_teenth(candidates)
    else:
        return find(candidates, schedule)

def find_teenth(candidates):
    for date in candidates:
        if date.day >= 13 and date.day <= 19:
            return date

def find(candidates, schedule):
    index = -1 if schedule == 'last' else int(schedule[0]) - 1
    return candidates[index]

